File: AUGUST2020/31.py
def inorder(k):
        while(k.left!=None):
            k=k.left
        return k
    
#important for when we have only left child.
#so we need to take the largest of them.
    
def inorder1(k):
    while(k.right!=None):
            k=k.right
    return k
        
def delete(root,key):
    if root:
        if root.val==key:
            if not(root.left) and not(root.right):
                return None
            if root.right:
                    j=inorder(root.right)
                    root.val=j.val
                    root.right=delete(root.right,j.val)
                    return root
            else:
                    j=inorder1(root.left)
                    root.val=j.val
                    root.left=delete(root.left,j.val)
                    return root
        elif root.val<key:
            root.right=delete(root.right,key)
            return root
        else:
            root.left=delete(root.left,key)
            return root
    return None
                
class Solution:
    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        root=delete(root,key)
        return root

File: AUGUST2020/31.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
typedef struct TreeNode tr;

int ino(tr* node){
    while(node && node->left)
        node=node->left;
    return node->val;
}

int ino1(tr* node){
    while(node && node->right)
        node=node->right;
    return node->val;
}

tr* delete(tr* root, int key){
    if (root){
        if (root->val==key){
            if (!(root->left) && !(root->right))
                return NULL;
            else if (root->right){
                int j=ino(root->right);
                root->val=j;
                root->right=delete(root->right,j);
                return root;
            }
            else{
                int j=ino1(root->left);
                root->val=j;
                root->left=delete(root->left,j);
                return root;
            }
        }
        else if (root->val<key){
            root->right=delete(root->right,key);
            return root;}
        else{
            root->left=delete(root->left,key);
            return root;
        }   
    }
    return NULL;
}

struct TreeNode* deleteNode(struct TreeNode* root, int key){
    root=delete(root,key);
    return root;
}
